tensorboard/webapp/widgets/line_chart_v2/data_transformer.ts

```diff
@@ -0,0 +1,99 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {DataSeries} from './lib/public_types';
+
+/**
+ * Smoothes data series in y axis using smoothing algorithm from classical TensorBoard
+ * circa 2019-2020. 1st-order IIR low-pass filter to attenuate the higher-frequency
+ * components of the time-series.
+ * @param data DataSeries to smooth
+ * @param smoothingWeight Degree of smoothing. Number between 0 and 1, inclusive.
+ */
+export async function classicSmoothing(
+  data: DataSeries[],
+  smoothingWeight: number
+): Promise<Array<{srcId: string; points: DataSeries['points']}>> {
+  if (!data.length) [];
+
+  if (!Number.isFinite(smoothingWeight)) {
+    smoothingWeight = 0;
+  }
+  smoothingWeight = Math.max(0, Math.min(smoothingWeight, 1));
+
+  const results: Array<{srcId: string; points: DataSeries['points']}> = [];
+
+  for (const series of data) {
+    if (!series.points.length) {
+      results.push({
+        srcId: series.id,
+        points: [],
+      });
+      continue;
+    }
+
+    let last = series.points.length > 0 ? 0 : NaN;
+    let numAccum = 0;
+
+    const initialYVal = series.points[0].y;
+    const isConstant = series.points.every((point) => point.y == initialYVal);
```

Comment:
Optional, but maybe consider extracting `smoothSeries(series.points)`?
imho makes it easier to see the nested loop structure; I misread it the
first time and thought that these `isConstant`s were being computed for
every point, which is not the case (good).

Reply:
I am a bit leaning towards keeping the signature as is; while I do not have a proof handy, I suspect transforming data will be a major area of work later and will likely be a hot path for performance. When such time comes, we should really look into WASM or other ways of doing the transformation and, in such case, asynchronous function that operates in a larger batch (entire DataSeries instead of one series) would be beneficial (to avoid frequent shuffling of data between host RAM to VRAM).